File: Proyecto_Python/src/algoritmos/V_4_Parallel_Block.py
from concurrent.futures import ThreadPoolExecutor

def block_multiply(matrizA, matrizB, matrizC, i1, size1, size2):
    for j1 in range(0, size1, size2):
        for k1 in range(0, size1, size2):
            for i in range(i1, min(i1 + size2, size1)):
                for j in range(j1, min(j1 + size2, size1)):
                    for k in range(k1, min(k1 + size2, size1)):
                        matrizC[k][i] += matrizA[k][j] * matrizB[j][i]

def alg_V_4_ParallelBlockTres(matrizA, matrizB, size1, size2):
    matrizC = [[0.0 for _ in range(size1)] for _ in range(size1)]
    num_blocks = (size1 + size2 - 1) // size2  # Calculate the number of blocks
    with ThreadPoolExecutor() as executor:
        # Dispatch each block to a separate thread
        futures = [executor.submit(block_multiply, matrizA, matrizB, matrizC, i1, size1, size2)
                for i1 in range(0, size1, size2)]
        for future in futures:
            future.result()  # Wait for all threads to complete
    return matrizC

def multiply(matrizA, matrizB):
    N = len(matrizA)
    P = len(matrizB[0])
    alg_V_4_ParallelBlockTres(matrizA, matrizB, N, P)

File: src/V_4_Parallel_Block.java
import java.util.stream.IntStream;

public class V_4_Parallel_Block {
    public void alg_V_4_ParallelBlockTres(double[][] matrizA, double[][] matrizB, int size1, int size2) {
        double[][] matrizC = new double[size1][size1];
        IntStream.range(0, 1).parallel().forEach(_i -> {
            for (int i1 = 0; i1 < size1; i1 += size2) {
                for (int j1 = 0; j1 < size1; j1 += size2) {
                    for (int k1 = 0; k1 < size1; k1 += size2) {
                        for (int i = i1; i < i1 + size2 && i < size1; i++) {
                            for (int j = j1; j < j1 + size2 && j < size1; j++) {
                                for (int k = k1; k < k1 + size2 && k < size1; k++) {
                                    matrizC[k][i] += matrizA[k][j] * matrizB[j][i];
                                }
                            }
                        }
                    }
                }
            }
        });
    }
}
